syncer/find_fork.go

```diff
@@ -0,0 +1,318 @@
+package syncer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/common/util"
+	"github.com/spacemeshos/go-spacemesh/fetch"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/layers"
+)
+
+var ErrPeerMeshChangedMidSession = errors.New("peer mesh changed mid session")
+
+type layerHash struct {
+	lid     types.LayerID
+	hash    types.Hash32
+	created time.Time
+}
+
+func newLayerHash(lid types.LayerID, hash types.Hash32) *layerHash {
+	return &layerHash{
+		lid:  lid,
+		hash: hash,
+	}
+}
+
+// boundary is used to define the to and from layers in the hash mesh requests to peers.
+// the hashes of the boundary layers are known to the node and are used to double-check that
+// the peer has not changed its opinions on those layers.
+// if the boundary hashes change during the fork-finding session, the session is aborted.
+type boundary [2]*layerHash
+
+func newBoundary() *boundary {
+	return &boundary{
+		newLayerHash(types.LayerID{}, types.Hash32{}),
+		newLayerHash(types.LayerID{}, types.Hash32{}),
+	}
+}
+
+func (b *boundary) setFrom(lh *layerHash) {
+	b[0] = lh
+}
+
+func (b *boundary) setTo(lh *layerHash) {
+	b[1] = lh
+}
+
+func (b *boundary) fromLayer() types.LayerID {
+	return b[0].lid
+}
+
+func (b *boundary) fomHash() types.Hash32 {
+	return b[0].hash
+}
+
+func (b *boundary) toLayer() types.LayerID {
+	return b[1].lid
+}
+
+func (b *boundary) toHash() types.Hash32 {
+	return b[1].hash
+}
+
+func (b *boundary) MarshalLogObject(encoder log.ObjectEncoder) error {
+	encoder.AddUint32("from", b.fromLayer().Uint32())
+	encoder.AddString("from_hash", b.fomHash().ShortString())
+	encoder.AddUint32("to", b.toLayer().Uint32())
+	encoder.AddString("to_hash", b.toHash().ShortString())
+	return nil
+}
+
+type ForkFinder struct {
+	logger           log.Log
+	db               sql.Executor
+	fetcher          fetcher
+	maxHashesInReq   uint32
+	maxStaleDuration time.Duration
+
+	mu          sync.Mutex
+	agreedPeers map[p2p.Peer]*layerHash
+}
+
+func NewForkFinder(lg log.Log, db sql.Executor, f fetcher, maxHashes uint32, maxStale time.Duration) *ForkFinder {
+	return &ForkFinder{
+		logger:           lg,
+		db:               db,
+		fetcher:          f,
+		maxHashesInReq:   maxHashes,
+		maxStaleDuration: maxStale,
+		agreedPeers:      make(map[p2p.Peer]*layerHash),
+	}
+}
+
+// Purge cached agreements with peers.
+func (ff *ForkFinder) Purge(all bool, toPurge ...p2p.Peer) {
+	ff.mu.Lock()
+	defer ff.mu.Unlock()
+
+	if all {
+		ff.agreedPeers = make(map[p2p.Peer]*layerHash)
+		return
+	}
+
+	if len(toPurge) > 0 {
+		for _, p := range toPurge {
+			delete(ff.agreedPeers, p)
+		}
+		return
+	}
+
+	uniquePeers := util.SliceToSetStringer(ff.fetcher.GetPeers())
```

Comment:
it doesn't make sense to do it like that. peer.String() encodes Peer (which is string) using base58.  p2p.Peer can be used as a key